Release v1.20.0 (#1536)
heremaps/here-data-sdk-cpp: CHANGELOG.md

## v1.20.0 (06/08/2024)

**Common**
* Updated links and filenames in the documentation.
* Improved the CI scripts and pipelines.

**olp-cpp-sdk-authentication**

> * Fixed the DateTime parser to respect DST timezones on iOS.

Reviewer: ```suggestion
* Fixed the `DateTime` parser to respect DST timezones on iOS.
```

Author: Done

> * Fixed determining default credentials file absolute path on Windows for new distributions.

Reviewer: I'm not sure I understand what this means. I think you fixed the mechanism that determines the absolute path of the credentials file on Windows, correct? When you say "for new distributions", do you mean new WINDOWS or new SDK?

Author: Yes, we fixed the mechanism. The old version of the mechanism didn't work on the new Windows distributions, Like Windows 11.

Reviewer: OK, with that said, I propose to change this to: 
```suggestion
* Fixed the mechanism that determines the absolute path of the credentials file on new Windows versions, such as Windows 11.
```
I don't think I've seen the word "distribution" used in the Windows context so I went with "version".

Author: Done

**olp-cpp-sdk-core**

> * Extend logging with `olp::logging::LogContext` and `olp::logging::ScopedLogContext` to give users ability to manage logs better.

Reviewer: ```suggestion
* Extended logging with `olp::logging::LogContext` and `olp::logging::ScopedLogContext` to facilitate better management of logs.
```

Author: Done

> * Added logging contexts usage to `olp::http::NetworkCurl` class.

Reviewer: ```suggestion
* Added logging contexts usage to the `olp::http::NetworkCurl` class.
```

Author: Done

> * Added a new error code `olp::client::ErrorCode::NoSpaceLeft` to signal shortage of available space on the target device.

Reviewer: ```suggestion
* Added a new error code `olp::client::ErrorCode::NoSpaceLeft` to signal a shortage of available space on the target device.
```

Author: Done

> * To be able to propagate actual operation error, extended `olp::cache::KeyValueCache` and `olp::cache::DefaultCache` with methods `Read`, `Write`, `Delete` and `DeleteByPrefix`, which do the same actions as `Put`, `Get`, `Remove` and `RemoveKeysWithPrefix` correspondently, but return `olp::cache::OperationOutcome`.

Reviewer: ```suggestion
* Extended  `olp::cache::KeyValueCache` and `olp::cache::DefaultCache` with new methods: Read`, `Write`, `Delete`, and `DeleteByPrefix` to allow propagation of actual operation errors.
These methods correspond to `Put`, `Get`, `Remove`, and `RemoveKeysWithPrefix`, but return `olp::cache::OperationOutcome`.
```

> * Extended `olp::client::OlpClientSettings` with a new setting `propagate_all_cache_errors` to allow users enabling the propagation of all cache errors.

Reviewer: ```suggestion
* Extended `olp::client::OlpClientSettings` with a new setting `propagate_all_cache_errors` that allows to enable the propagation of all cache errors.
```

Author: Done

> * Added a new way to serialize a byte vector to JSON. 

Reviewer: Shouldn't we say what new way that is? Or maybe it isn't really possible to describe it in a couple of words?

Author: Rephrased it.

> * Updated the logic to use generated session ID for the background network session for iOS.

Reviewer: ```suggestion
* Updated the logic to use a generated session ID for the background network session on iOS.
```

Author: Done

> * Added support of Android-like build on desktop to be able to run tests on host PC.

Reviewer: ```suggestion
* Added support of Android-like build on desktops to allow running tests on host PCs.
```

Author: Done

* Fixed deprecation warnings related to C++17.
* Added background timeouts for iOS.
* Made CMake option `OLP_SDK_ENABLE_IOS_BACKGROUND_DOWNLOAD` usable only for iOS.

> * Added missing thread names for heavy operations to ease profiling and debugging.

Reviewer: "heavy operations" -> in what way heavy? Resource-heavy? CPU-intensive? Please clarify.

Author: I didn't specify a certain kind of operation because it implies all kinds of heavy operations, including CPU-intensive, RAM-heavy, disk-heavy, time-consuming etc.

Author: Done

**olp-cpp-sdk-dataservice-read**
* Optimized memory allocations.

> * Added `DeleteFromCache` methods to `olp::dataservice::read::VersionedLayerClient` and `olp::dataservice::read::VolatileLayerClient` to be able to propagate actual operation error.

Reviewer: ```suggestion
* Added a `DeleteFromCache` method to `olp::dataservice::read::VersionedLayerClient` and `olp::dataservice::read::VolatileLayerClient` to allow the propagation of the actual operation error.
```

Author: Done
